### PR TITLE
chore: user stricter types for Compass preferences

### PR DESCRIPTION
**Pull request by addaleax**
Make sure that types match what they are expected to match, and make sure that the Ampersand props definition contains exactly what it is supposed to contain.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [x] Misc

## Types of changes

- [ ] *Backport Needed*
- [x] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-settings/src/components/settings/privacy.tsx [hunk 1]
```diff
@@ -11,15 +11,11 @@ import {
...
 
 type PrivacySettingsProps = {
   handleChange: (field: PrivacyFields, value: boolean) => void;
-  autoUpdates?: boolean;
-  enableMaps?: boolean;
-  trackErrors?: boolean;
-  trackUsageStatistics?: boolean;
-  enableFeedbackPanel?: boolean;
-};
+} & Pick<Partial<UserConfigurablePreferences>, PrivacyFields>;
```
Comment (alenakhineika) [inline packages/compass-settings/src/components/settings/privacy.tsx:18]: 🔥

File: packages/compass-preferences-model/src/preferences.ts [hunk 2]
```diff
@@ -44,15 +60,31 @@ declare class PreferencesAmpersandModel {
...
+  ? 'object'
+  : never;
+
+type PreferenceDefinition<K extends keyof GlobalPreferences> = {
+  type: AmpersandType<GlobalPreferences[K]>;
+  default?: GlobalPreferences[K];
+  required: boolean;
+  ui: K extends keyof UserConfigurablePreferences ? true : false;
+  cli: K extends keyof InternalUserPreferences ? false : boolean;
+  global: K extends keyof InternalUserPreferences ? false : boolean;
```
[the remark this answers is not in the document]
Comment (addaleax) [reply packages/compass-preferences-model/src/preferences.ts:83]: > Shouldn't global and ci extend GlobalPreferences?

What does “extend” mean here?

> And how are those different from UI so they are `false : boolean` and not `true : false`?

The logic is that for `ui`, `UserConfigurablePreferences` is the sole source of truth for whether a preference should be configurable in the UI or not.

For `cli` and `global`, we don’t want users to be able to override properties that we consider internal to Compass, but if they are not internal, then that does not automatically mean that it makes sense to override them from CLI/config file.

File: packages/compass-preferences-model/src/preferences.ts [hunk 3]
```diff
@@ -84,9 +116,9 @@ const preferencesProps: {
     type: 'boolean',
     required: true,
     default: false,
-    ui: true,
-    cli: true,
-    global: true,
+    ui: false,
+    cli: false,
+    global: false,
```
[the remark this answers is not in the document]
Comment (addaleax) [reply packages/compass-preferences-model/src/preferences.ts:121]: I’m not sure if that’s okay from a legal point of view, so I’d also check with product. Even if we do, I think we should keep this preference internal, as a way of storing whether the settings modal has been displayed – we can add a second option with a better name if we think it’s a good idea to add one.

How about you add a suggestion to the table in the tech design & ping Julia on it?

File: packages/compass-preferences-model/src/preferences.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alenakhineika) [reply packages/compass-preferences-model/src/preferences.ts:83]: Got it!

**Review by alenakhineika — APPROVED** [review]

**Pull request merged**